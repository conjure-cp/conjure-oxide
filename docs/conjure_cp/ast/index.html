<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `ast` mod in crate `conjure_cp`."><title>conjure_cp::ast - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="conjure_cp" data-themes="" data-resource-suffix="" data-rustdoc-version="1.93.0-nightly (278a90913 2025-10-28)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module ast</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../conjure_cp/index.html">conjure_<wbr>cp</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module ast</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate conjure_<wbr>cp</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">conjure_cp</a></div><h1>Module <span>ast</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/conjure_cp_core/lib.rs.html#7">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="ac_operators/index.html" title="mod conjure_cp::ast::ac_operators">ac_<wbr>operators</a></dt><dt><a class="mod" href="categories/index.html" title="mod conjure_cp::ast::categories">categories</a></dt><dt><a class="mod" href="comprehension/index.html" title="mod conjure_cp::ast::comprehension">comprehension</a></dt><dt><a class="mod" href="declaration/index.html" title="mod conjure_cp::ast::declaration">declaration</a></dt><dt><a class="mod" href="matrix/index.html" title="mod conjure_cp::ast::matrix">matrix</a></dt><dd>Utility functions for working with matrices.</dd><dt><a class="mod" href="pretty/index.html" title="mod conjure_cp::ast::pretty">pretty</a></dt><dd>Functions for pretty printing Conjure models.</dd><dt><a class="mod" href="records/index.html" title="mod conjure_cp::ast::records">records</a></dt><dt><a class="mod" href="serde/index.html" title="mod conjure_cp::ast::serde">serde</a></dt><dd>Serde serialization/ deserialization helpers.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.CnfClause.html" title="struct conjure_cp::ast::CnfClause">CnfClause</a></dt><dt><a class="struct" href="struct.DecisionVariable.html" title="struct conjure_cp::ast::DecisionVariable">Decision<wbr>Variable</a></dt><dd>Represents a decision variable within a computational model.</dd><dt><a class="struct" href="struct.DeclarationPtr.html" title="struct conjure_cp::ast::DeclarationPtr">Declaration<wbr>Ptr</a></dt><dd>A shared pointer to a [<code>Declaration</code>].</dd><dt><a class="struct" href="struct.Metadata.html" title="struct conjure_cp::ast::Metadata">Metadata</a></dt><dt><a class="struct" href="struct.Model.html" title="struct conjure_cp::ast::Model">Model</a></dt><dd>An Essence model.</dd><dt><a class="struct" href="struct.Moo.html" title="struct conjure_cp::ast::Moo">Moo</a></dt><dd>A clone-on-write, reference counted pointer to an AST type.</dd><dt><a class="struct" href="struct.RecordEntry.html" title="struct conjure_cp::ast::RecordEntry">Record<wbr>Entry</a></dt><dt><a class="struct" href="struct.SerdeModel.html" title="struct conjure_cp::ast::SerdeModel">Serde<wbr>Model</a></dt><dd>A model that is de/serializable using <code>serde</code>.</dd><dt><a class="struct" href="struct.SubModel.html" title="struct conjure_cp::ast::SubModel">SubModel</a></dt><dd>A sub-model, representing a lexical scope in the model.</dd><dt><a class="struct" href="struct.SymbolTable.html" title="struct conjure_cp::ast::SymbolTable">Symbol<wbr>Table</a></dt><dd>The global symbol table, mapping names to their definitions.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="enum" href="enum.AbstractLiteral.html" title="enum conjure_cp::ast::AbstractLiteral">Abstract<wbr>Literal</a></dt><dt><a class="enum" href="enum.Atom.html" title="enum conjure_cp::ast::Atom">Atom</a></dt><dd>An <code>Atom</code> is an indivisible expression, such as a literal or a reference.</dd><dt><a class="enum" href="enum.DeclarationKind.html" title="enum conjure_cp::ast::DeclarationKind">Declaration<wbr>Kind</a></dt><dd>A specific kind of declaration.</dd><dt><a class="enum" href="enum.Domain.html" title="enum conjure_cp::ast::Domain">Domain</a></dt><dt><a class="enum" href="enum.DomainOpError.html" title="enum conjure_cp::ast::DomainOpError">Domain<wbr>OpError</a></dt><dd>An error thrown by an operation on domains.</dd><dt><a class="enum" href="enum.Expression.html" title="enum conjure_cp::ast::Expression">Expression</a></dt><dd>Represents different types of expressions used to define rules and constraints in the model.</dd><dt><a class="enum" href="enum.Literal.html" title="enum conjure_cp::ast::Literal">Literal</a></dt><dd>A literal value, equivalent to constants in Conjure.</dd><dt><a class="enum" href="enum.Name.html" title="enum conjure_cp::ast::Name">Name</a></dt><dd>A reference to an object stored in the [<code>SymbolTable</code>].</dd><dt><a class="enum" href="enum.Range.html" title="enum conjure_cp::ast::Range">Range</a></dt><dt><a class="enum" href="enum.ReturnType.html" title="enum conjure_cp::ast::ReturnType">Return<wbr>Type</a></dt><dt><a class="enum" href="enum.SetAttr.html" title="enum conjure_cp::ast::SetAttr">SetAttr</a></dt></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.Typeable.html" title="trait conjure_cp::ast::Typeable">Typeable</a></dt><dd>Something with a return type</dd></dl></section></div></main></body></html>