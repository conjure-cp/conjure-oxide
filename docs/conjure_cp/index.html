<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `conjure_cp` crate."><title>conjure_cp - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="conjure_cp" data-themes="" data-resource-suffix="" data-rustdoc-version="1.93.0-nightly (278a90913 2025-10-28)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Crate conjure_cp</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../conjure_cp/index.html">conjure_<wbr>cp</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#modules">Crate Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>conjure_<wbr>cp</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/conjure_cp/lib.rs.html#1-13">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="ast/index.html" title="mod conjure_cp::ast">ast</a></dt><dt><a class="mod" href="context/index.html" title="mod conjure_cp::context">context</a></dt><dt><a class="mod" href="defaults/index.html" title="mod conjure_cp::defaults">defaults</a></dt><dt><a class="mod" href="error/index.html" title="mod conjure_cp::error">error</a></dt><dd>Top-level error types for Conjure-Oxide.</dd><dt><a class="mod" href="parse/index.html" title="mod conjure_cp::parse">parse</a></dt><dd>Essence parsers.</dd><dt><a class="mod" href="representation/index.html" title="mod conjure_cp::representation">representation</a></dt><dt><a class="mod" href="rule_engine/index.html" title="mod conjure_cp::rule_engine">rule_<wbr>engine</a></dt><dt><a class="mod" href="solver/index.html" title="mod conjure_cp::solver">solver</a></dt><dd>A high-level API for interacting with constraints solvers.</dd><dt><a class="mod" href="stats/index.html" title="mod conjure_cp::stats">stats</a></dt></dl><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="macro" href="macro.bug.html" title="macro conjure_cp::bug">bug</a></dt><dd>Triggers a panic with a detailed bug report message, while ensuring the panic is ignored in coverage reports.</dd><dt><a class="macro" href="macro.domain_int.html" title="macro conjure_cp::domain_int">domain_<wbr>int</a></dt><dd>Creates a <a href="ast/enum.Domain.html#variant.Int" title="variant conjure_cp::ast::Domain::Int"><code>Domain::Int</code></a>.</dd><dt><a class="macro" href="macro.error.html" title="macro conjure_cp::error">error</a></dt><dt><a class="macro" href="macro.essence_expr.html" title="macro conjure_cp::essence_expr">essence_<wbr>expr</a></dt><dd>Parses an Essence expression into its corresponding Conjure AST at compile time.</dd><dt><a class="macro" href="macro.essence_vec.html" title="macro conjure_cp::essence_vec">essence_<wbr>vec</a></dt><dd>Parses a sequence of Essence expressions into a vector of Conjure AST instances</dd><dt><a class="macro" href="macro.into_matrix.html" title="macro conjure_cp::into_matrix">into_<wbr>matrix</a></dt><dd>Creates a new matrix <a href="ast/enum.AbstractLiteral.html" title="enum conjure_cp::ast::AbstractLiteral"><code>AbstractLiteral</code></a> from some <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>, optionally with some index domain.</dd><dt><a class="macro" href="macro.into_matrix_expr.html" title="macro conjure_cp::into_matrix_expr">into_<wbr>matrix_<wbr>expr</a></dt><dd>Creates a new matrix as an <a href="ast/enum.Expression.html" title="enum conjure_cp::ast::Expression"><code>Expression</code></a> from a (Rust) vector, optionally with some index
domain.</dd><dt><a class="macro" href="macro.matrix.html" title="macro conjure_cp::matrix">matrix</a></dt><dd>Creates a new matrix <a href="ast/enum.AbstractLiteral.html" title="enum conjure_cp::ast::AbstractLiteral"><code>AbstractLiteral</code></a> optionally with some index domain.</dd><dt><a class="macro" href="macro.matrix_expr.html" title="macro conjure_cp::matrix_expr">matrix_<wbr>expr</a></dt><dd>Creates a new matrix as an <a href="ast/enum.Expression.html" title="enum conjure_cp::ast::Expression"><code>Expression</code></a>, optionally with some index domain.</dd><dt><a class="macro" href="macro.range.html" title="macro conjure_cp::range">range</a></dt><dd>Creates a <a href="ast/enum.Range.html" title="enum conjure_cp::ast::Range"><code>Range</code></a>.</dd><dt><a class="macro" href="macro.register_representation.html" title="macro conjure_cp::register_representation">register_<wbr>representation</a></dt><dt><a class="macro" href="macro.throw_error.html" title="macro conjure_cp::throw_error">throw_<wbr>error</a></dt></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.Model.html" title="struct conjure_cp::Model">Model</a></dt><dd>An Essence model.</dd></dl></section></div></main></body></html>