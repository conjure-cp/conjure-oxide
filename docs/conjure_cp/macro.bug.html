<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Triggers a panic with a detailed bug report message, while ensuring the panic is ignored in coverage reports."><title>bug in conjure_cp - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="conjure_cp" data-themes="" data-resource-suffix="" data-rustdoc-version="1.93.0-nightly (278a90913 2025-10-28)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">bug</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../conjure_cp/index.html">conjure_<wbr>cp</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">bug</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#parameters" title="Parameters">Parameters</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate conjure_<wbr>cp</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">conjure_cp</a></div><h1>Macro <span class="macro">bug</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/conjure_cp_core/bug.rs.html#11">Source</a> </span></div><pre class="rust item-decl"><code>macro_rules! bug {
    ($msg:expr) =&gt; { ... };
    ($msg:expr, $($arg:tt)*) =&gt; { ... };
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Triggers a panic with a detailed bug report message, while ensuring the panic is ignored in coverage reports.</p>
<p>This macro is useful in situations where an unreachable code path is hit or when a bug occurs.</p>
<h2 id="parameters"><a class="doc-anchor" href="#parameters">ยง</a>Parameters</h2>
<ul>
<li><code>msg</code>: A string expression describing the cause of the panic or bug.</li>
</ul>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code></code></pre></div></div></details></section></div></main></body></html>