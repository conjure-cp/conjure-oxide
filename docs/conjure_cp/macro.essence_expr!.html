<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.essence_expr.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.essence_expr.html">macro.essence_expr.html</a>...</p>
    <script>location.replace("macro.essence_expr.html" + location.search + location.hash);</script>
</body>
</html>