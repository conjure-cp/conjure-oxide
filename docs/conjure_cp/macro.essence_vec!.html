<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.essence_vec.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.essence_vec.html">macro.essence_vec.html</a>...</p>
    <script>location.replace("macro.essence_vec.html" + location.search + location.hash);</script>
</body>
</html>