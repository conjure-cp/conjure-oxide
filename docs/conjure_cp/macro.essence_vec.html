<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Parses a sequence of Essence expressions into a vector of Conjure AST instances"><title>essence_vec in conjure_cp - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="conjure_cp" data-themes="" data-resource-suffix="" data-rustdoc-version="1.93.0-nightly (278a90913 2025-10-28)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">essence_vec</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../conjure_cp/index.html">conjure_<wbr>cp</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">essence_<wbr>vec</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#example" title="Example">Example</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate conjure_<wbr>cp</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">conjure_cp</a></div><h1>Macro <span class="macro">essence_<wbr>vec</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/conjure_cp_essence_macros/lib.rs.html#88">Source</a> </span></div><pre class="rust item-decl"><code>essence_vec!() { <span class="comment">/* proc-macro */</span> }</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Parses a sequence of Essence expressions into a vector of Conjure AST instances</p>
<h3 id="example"><a class="doc-anchor" href="#example">ยง</a>Example</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>conjure_cp::ast::{Atom, Expression, Moo, Metadata};
<span class="kw">use </span>conjure_cp::matrix_expr;
<span class="kw">use </span>conjure_cp_essence_macros::essence_vec;

<span class="kw">let </span>exprs = <span class="macro">essence_vec!</span>(<span class="number">2 </span>+ <span class="number">2</span>, <span class="bool-val">false </span>= <span class="bool-val">true</span>);
<span class="macro">println!</span>(<span class="string">"{:?}"</span>, exprs);
<span class="macro">assert_eq!</span>(exprs.len(), <span class="number">2</span>);
<span class="macro">assert_eq!</span>(
    exprs[<span class="number">0</span>],
    Expression::Sum(Metadata::new(), Moo::new(<span class="macro">matrix_expr!</span>[
        Expression::Atomic(Metadata::new(), <span class="number">2</span>.into()),
        Expression::Atomic(Metadata::new(), <span class="number">2</span>.into())
    ]))
);
<span class="macro">assert_eq!</span>(
   exprs[<span class="number">1</span>],
    Expression::Eq(Metadata::new(),
        Moo::new(Expression::Atomic(Metadata::new(), <span class="bool-val">false</span>.into())),
        Moo::new(Expression::Atomic(Metadata::new(), <span class="bool-val">true</span>.into()))
    )
);</code></pre></div></div></details></section></div></main></body></html>