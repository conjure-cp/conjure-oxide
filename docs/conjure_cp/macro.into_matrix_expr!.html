<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.into_matrix_expr.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.into_matrix_expr.html">macro.into_matrix_expr.html</a>...</p>
    <script>location.replace("macro.into_matrix_expr.html" + location.search + location.hash);</script>
</body>
</html>