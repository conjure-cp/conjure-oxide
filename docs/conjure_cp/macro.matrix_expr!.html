<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.matrix_expr.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.matrix_expr.html">macro.matrix_expr.html</a>...</p>
    <script>location.replace("macro.matrix_expr.html" + location.search + location.hash);</script>
</body>
</html>