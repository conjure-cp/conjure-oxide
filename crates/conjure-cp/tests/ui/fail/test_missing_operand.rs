use conjure_cp::essence_expr;

fn main() {
    essence_expr!(2 + (3 /));
}