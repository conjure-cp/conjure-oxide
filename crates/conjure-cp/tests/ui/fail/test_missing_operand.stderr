error: Syntax error: Expected an integer here
       2 + (3 /)
                ^-- Error here
 --> tests/ui/fail/test_missing_operand.rs:4:5
  |
4 |     essence_expr!(2 + (3 /));
  |     ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `essence_expr` (in Nightly builds, run with -Z macro-backtrace for more info)
